DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces_overview/csr-pipes/src/vector_add.cpp

```diff
@@ -0,0 +1,124 @@
+#include <iostream>
+
+// oneAPI headers
+#include <sycl/sycl.hpp>
+#include <sycl/ext/intel/fpga_extensions.hpp>
+#include "exception_handler.hpp"
+
+// Forward declare the kernel name in the global scope. This is an FPGA best
+// practice that reduces name mangling in the optimization reports.
+class SimpleVAddPipes;
+
+// Forward declare pipe names to reduce name mangling (maybe get rid of; pipes
+// already have bad names)
+class IDPipeA;
+class IDPipeB;
+class IDPipeC;
+
+#define VECT_SIZE 256
+
+using PipeProps = decltype(sycl::ext::oneapi::experimental::properties(
+    sycl::ext::intel::experimental::ready_latency<0>));
+
+using InputPipeA =
+    sycl::ext::intel::experimental::pipe<IDPipeA, int, 0,
+                                         PipeProps>;
+using InputPipeB =
+    sycl::ext::intel::experimental::pipe<IDPipeB, int, 0,
+                                         PipeProps>;
+
+using CSRPipeProps = decltype(sycl::ext::oneapi::experimental::properties(
+    sycl::ext::intel::experimental::protocol_avalon_mm));
+
+// this csr pipe will only be read and write once with the total sum
+using OutputPipeC =
+    sycl::ext::intel::experimental::pipe<IDPipeC, int, 0,
+                                         CSRPipeProps>;
+
+struct SimpleVAddKernelPipes {
+  int len;
+
+  void operator()() const {
+    int sum_total = 0;
+    for (int idx = 0; idx < len; idx++) {
+      int a_val = InputPipeA::read();
+      int b_val = InputPipeB::read();
+      int sum = a_val + b_val;
+
+      sum_total += sum;
+    }
+
+    // Write to OutputPipeC only once per kernel invocation
+    OutputPipeC::write(sum_total);
+  }
+};
+
+int main() {
+  try {
+    // Use compile-time macros to select either:
+    //  - the FPGA emulator device (CPU emulation of the FPGA)
+    //  - the FPGA device (a real FPGA)
+    //  - the simulator device
+#if FPGA_SIMULATOR
+    auto selector = sycl::ext::intel::fpga_simulator_selector_v;
+#elif FPGA_HARDWARE
+    auto selector = sycl::ext::intel::fpga_selector_v;
+#else  // #if FPGA_EMULATOR
+    auto selector = sycl::ext::intel::fpga_emulator_selector_v;
+#endif
+
+    // create the device queue
+    sycl::queue q(selector, fpga_tools::exception_handler);
+
+    int count = VECT_SIZE;  // pass array size by value
+
+    int expectedSum = 0;
```

Comment:
snake_case for variables
```suggestion
    int expected_sum = 0;
```